### PR TITLE
ReflectionBasedDCS: Fix 5 test failures, MemberInfo, DateTimeOffset, KeyValuePair

### PR DESCRIPTION
**Pull request by khdang**
Fix the following 5 tests:

DCS_DateTimeOffsetAsRoot
DCS_DataMemberAttribute
DCS_DuplicatedKnownTypesWithAdapterThroughConstructor
DCS_DuplicatedKeyDateTimeOffset
DCS_MyPersonSurrogate

cc: @SGuyGe @shmao @zhenlan

### CODE REVIEW

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionXmlFormatReader.cs [hunk 1]
```diff
@@ -43,6 +43,16 @@ internal object ReflectionReadClass(XmlReaderDelegator xmlReader, XmlObjectSeria
             object obj = ReflectionCreateObject(_classContract);
             context.AddNewObject(obj);
             ReflectionReadMembers(obj, xmlReader, context, memberNames, memberNamespaces);
+
+            if (obj.GetType() == typeof(DateTimeOffsetAdapter))
```
Comment (SGuyGe) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionXmlFormatReader.cs]: We should not allow concrete adapters to leak into ReflectionXmlFormatReader or ReflectionXmlFormatWriter.

Comment (khdang) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionXmlFormatReader.cs]: I've created https://github.com/dotnet/corefx/issues/8507. We can work on that after ReflectionBasedDCS change is done.

**Comment by khdang** [comment]
If there is no other concerns I'll merge this PR by end of day. Thanks.

**Comment by SGuyGe** [comment]
LGTM

**Pull request merged**